util/filesort/filesort.go

```diff
@@ -0,0 +1,471 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filesort
+
+import (
+	"container/heap"
+	"encoding/binary"
+	"io"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type comparableRow struct {
+	key    []types.Datum
+	val    []types.Datum
+	handle int64
+}
+
+type item struct {
+	index int // source file index
+	value *comparableRow
+}
+
+// Min-heap of comparableRows
+type rowHeap struct {
+	sc     *variable.StatementContext
+	ims    []*item
+	byDesc []bool
+	err    error
+}
+
+func lessThan(sc *variable.StatementContext, i []types.Datum, j []types.Datum, byDesc []bool) (bool, error) {
+	for k := range byDesc {
+		v1 := i[k]
+		v2 := j[k]
+
+		ret, err := v1.CompareDatum(sc, v2)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+
+		if byDesc[k] {
+			ret = -ret
+		}
+
+		if ret < 0 {
+			return true, nil
+		} else if ret > 0 {
+			return false, nil
+		}
+	}
+	return false, nil
+}
+
+func (rh *rowHeap) Len() int { return len(rh.ims) }
+
+func (rh *rowHeap) Swap(i, j int) { rh.ims[i], rh.ims[j] = rh.ims[j], rh.ims[i] }
+
+func (rh *rowHeap) Less(i, j int) bool {
+	l := rh.ims[i].value.key
+	r := rh.ims[j].value.key
+	ret, err := lessThan(rh.sc, l, r, rh.byDesc)
+	if rh.err == nil {
+		rh.err = err
+	}
+	return ret
+}
+
+func (rh *rowHeap) Push(x interface{}) {
+	rh.ims = append(rh.ims, x.(*item))
+}
+
+func (rh *rowHeap) Pop() interface{} {
+	old := rh.ims
+	n := len(old)
+	x := old[n-1]
+	rh.ims = old[0 : n-1]
+	return x
+}
+
+// FileSorter sorts the given rows according to the byDesc order.
+// FileSorter can sort rows that exceed predefined memory capacity.
+type FileSorter struct {
+	keySize   int                        // size of key slice
+	valSize   int                        // size of val slice
+	bufSize   int                        // size of buf slice
+	tmpDir    string                     // working directory for file sort
+	sc        *variable.StatementContext // required by Datum comparison
+	buf       []*comparableRow           // in-momory buffer of rows
+	files     []string                   // files generated by file sort
+	byDesc    []bool                     // whether or not the specific column sorted in descending order
+	closed    bool
+	fetched   bool
+	rowHeap   *rowHeap
+	fds       []*os.File
+	fileCount int
+	err       error
+}
+
+// Builder builds a new FileSorter.
+type Builder struct {
+	sc      *variable.StatementContext
+	keySize int
+	valSize int
+	bufSize int
+	byDesc  []bool
+	tmpDir  string
+}
+
+// NewBuilder creates a new Builder.
+func NewBuilder() *Builder {
```

Comment:
@shenli Yes, you're right. I simply follow the convention used in `infoschema/builder.go`. Forget to notice that `NewBuilder` is required in that case since it needs to create unexported `infoSchema` structure. I'll fix it. Thanks!